src/peft/tuners/ia3.py

```diff
@@ -0,0 +1,542 @@
+        kwargs = {
+            "fan_in_fan_out": ia3_config.fan_in_fan_out,
+            "init_ia3_weights": ia3_config.init_ia3_weights,
+        }
+        key_list = [key for key, _ in self.model.named_modules()]
+        for key in key_list:
+            if isinstance(ia3_config.target_modules, str):
+                target_module_found = re.fullmatch(ia3_config.target_modules, key)
+            else:
+                target_module_found = any(
+                    self._is_valid_match(key, target_key) for target_key in ia3_config.target_modules
+                )
+            if target_module_found:
+                if isinstance(ia3_config.feedforward_modules, str):
+                    is_feedforward = re.fullmatch(ia3_config.feedforward_modules, key)
+                else:
+                    is_feedforward = any(key.endswith(target_key) for target_key in ia3_config.feedforward_modules)
+                if not is_target_modules_in_base_model:
+                    is_target_modules_in_base_model = True
+                parent, target, target_name = _get_submodules(self.model, key)
+                bias = target.bias is not None
+                if isinstance(target, IA3Layer):
+                    target.update_layer(
+                        adapter_name,
+                        ia3_config.init_ia3_weights,
+                    )
+                else:
+                    if loaded_in_8bit and isinstance(target, bnb.nn.Linear8bitLt):
+                        eightbit_kwargs = kwargs.copy()
+                        eightbit_kwargs.update(
+                            {
+                                "has_fp16_weights": target.state.has_fp16_weights,
+                                "memory_efficient_backward": target.state.memory_efficient_backward,
+                                "threshold": target.state.threshold,
+                                "index": target.index,
+                            }
+                        )
+                        new_module = Linear8bitLt(
+                            adapter_name,
+                            target.in_features,
+                            target.out_features,
+                            is_feedforward,
+                            bias=bias,
+                            **eightbit_kwargs,
+                        )
+                    else:
+                        if isinstance(target, torch.nn.Linear):
+                            in_features, out_features = target.in_features, target.out_features
+                            if kwargs["fan_in_fan_out"]:
+                                warnings.warn(
+                                    "fan_in_fan_out is set to True but the target module is `torch.nn.Linear`. "
+                                    "Setting fan_in_fan_out to False."
+                                )
+                                kwargs["fan_in_fan_out"] = ia3_config.fan_in_fan_out = False
+                        elif isinstance(target, Conv1D):
```

Comment:
Yes! This works for any Conv1D layer. In fact, this part of the code is borrowed from the LoRA implementation.

Reply:
Good point about LoRA already doing the same. I was a bit confused because, for instance, [this post](https://stackoverflow.com/a/56685503) implies that Conv1D is only equivalent for kernel size of 1.

Reply:
Oh my bad, I understood your question now. You're completely right, this is not for _any_ Conv1D layer, in terms of the general 1D convolution layer. This is only for instances of Conv1D in `transformers.pytorch_utils` , as implemented here: https://github.com/huggingface/transformers/blob/9a5d468ba0562e2d5edf9da787881fa227132bca/src/transformers/pytorch_utils.py#L85.  This is equivalent and also implemented as a matrix multiply.

 I was thinking only in terms of the usage of this "Conv1D" in `transformers.pytorch_utils`  (which is a `kernel_size`=1,` stride`=1 version of the general Conv1D in PyTorch here: https://github.com/pytorch/pytorch/blob/d0e5c681f5491fc5bceb3a7f18830f7c7437d7d2/torch/nn/modules/conv.py#L180)

Reply:
Oh I see, thanks a lot for the explanation. I automatically assumed that `Conv1D` was from PyTorch, I should have checked first. Now it makes sense to me.